src/ray/raylet/node_manager.cc

```diff
@@ -638,8 +639,22 @@ void NodeManager::ProcessGetTaskMessage(
 
 void NodeManager::ProcessDisconnectClientMessage(
     const std::shared_ptr<LocalClientConnection> &client) {
-  // Remove the dead worker from the pool and stop listening for messages.
   const std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
+  const std::shared_ptr<Worker> driver = worker_pool_.GetRegisteredDriver(client);
+  // This client can't be a worker and a driver.
+  RAY_CHECK(worker == nullptr || driver == nullptr);
+
+  // If both worker and driver are null, then this method has already been
```

Comment:
why is this true?

Reply:
This PR changes behavior so that `ProcessDisconnectClientMessage` can be called multiple times. When this method is called the first time, it will remove the worker/driver from `worker_pool_` via the call to

```cpp
worker_pool_.DisconnectWorker(worker);
```

or 

```cpp
worker_pool_.DisconnectDriver(driver);
```